### PR TITLE
Controller resource, support Controller in HA

### PR DESCRIPTION
**Pull request by danijeel**
New resource,  `appgatesdp_appliance_controller_activation`. This will allow us to enable controller functionallity on a pre-existing, seeded appliance instance.

While reviewing, start with reading the `*.markdown` files before the code for extra context.

Acceptance tests has been omitted since we have not yet created the required testing infrastructure for this, however I've thoroughly tested this manually, and created an example usage here: 

- https://github.com/appgate/sdp-tf-reference-architecture/tree/high-availability-controllers/deployment/aws/high-availability-controllers

This PR will solve the problems explained in https://github.com/appgate/terraform-provider-appgatesdp/issues/188

acceptance test on 5.4.2-25749-release

<details>

```bash
> make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./appgate -v -count 1 -parallel 20  -timeout 120m
=== RUN   TestNewClient
--- PASS: TestNewClient (0.00s)
=== RUN   TestLoginInternalServerError
--- PASS: TestLoginInternalServerError (0.00s)
=== RUN   TestConfigGetToken
=== RUN   TestConfigGetToken/test_before_5.4
2021/11/12 17:01:29 [DEBUG] Login OK
=== RUN   TestConfigGetToken/test_5.4_login
2021/11/12 17:01:29 [DEBUG] Login OK
=== RUN   TestConfigGetToken/invalid_client_version
=== RUN   TestConfigGetToken/500_login_response
2021/11/12 17:01:29 [DEBUG] Login failed, controller not responding, got HTTP 500
2021/11/12 17:01:30 [DEBUG] Login failed, controller not responding, got HTTP 500
=== RUN   TestConfigGetToken/502_login_response
2021/11/12 17:01:30 [DEBUG] Login failed, controller not responding, got HTTP 502
2021/11/12 17:01:31 [DEBUG] Login failed, controller not responding, got HTTP 502
=== RUN   TestConfigGetToken/503_login_response
2021/11/12 17:01:31 [DEBUG] Login failed, controller not responding, got HTTP 503
2021/11/12 17:01:31 [DEBUG] Login failed, controller not responding, got HTTP 503
=== RUN   TestConfigGetToken/406_login_response
--- PASS: TestConfigGetToken (1.61s)
    --- PASS: TestConfigGetToken/test_before_5.4 (0.00s)
    --- PASS: TestConfigGetToken/test_5.4_login (0.00s)
    --- PASS: TestConfigGetToken/invalid_client_version (0.00s)
    --- PASS: TestConfigGetToken/500_login_response (0.55s)
    --- PASS: TestConfigGetToken/502_login_response (0.59s)
    --- PASS: TestConfigGetToken/503_login_response (0.46s)
    --- PASS: TestConfigGetToken/406_login_response (0.00s)
=== RUN   TestConfigValidate
=== RUN   TestConfigValidate/ok_config_minimum_required
=== RUN   TestConfigValidate/invalid_appgate_URL
=== RUN   TestConfigValidate/invalid_token
=== RUN   TestConfigValidate/base64_token
=== RUN   TestConfigValidate/invalid_client_version
=== RUN   TestConfigValidate/invalid_username_password
--- PASS: TestConfigValidate (0.00s)
    --- PASS: TestConfigValidate/ok_config_minimum_required (0.00s)
    --- PASS: TestConfigValidate/invalid_appgate_URL (0.00s)
    --- PASS: TestConfigValidate/invalid_token (0.00s)
    --- PASS: TestConfigValidate/base64_token (0.00s)
    --- PASS: TestConfigValidate/invalid_client_version (0.00s)
    --- PASS: TestConfigValidate/invalid_username_password (0.00s)
=== RUN   TestAccAppgateAdministrativeRoleDataSource
=== PAUSE TestAccAppgateAdministrativeRoleDataSource
=== RUN   TestAccAppgateApplianceCustomizationDataSource
=== PAUSE TestAccAppgateApplianceCustomizationDataSource
=== RUN   TestAccAppgateApplianceSeedDataSource
--- PASS: TestAccAppgateApplianceSeedDataSource (8.47s)
=== RUN   TestAccAppgateCADataSource
--- PASS: TestAccAppgateCADataSource (1.93s)
=== RUN   TestAccAppgateConditionDataSource
--- PASS: TestAccAppgateConditionDataSource (2.67s)
=== RUN   TestAccAppgateEntitlementDataSource
--- PASS: TestAccAppgateEntitlementDataSource (12.90s)
=== RUN   TestAccAppgateGlobalSettingsDataSource
--- PASS: TestAccAppgateGlobalSettingsDataSource (2.57s)
=== RUN   TestAccAppgateIdentityProviderDataSource
--- PASS: TestAccAppgateIdentityProviderDataSource (2.52s)
=== RUN   TestAccAppgateIPPoolDataSource
--- PASS: TestAccAppgateIPPoolDataSource (2.57s)
=== RUN   TestAccAppgateLocalUserDataSource
--- PASS: TestAccAppgateLocalUserDataSource (2.70s)
=== RUN   TestAccAppgateMfaProviderDataSource
=== PAUSE TestAccAppgateMfaProviderDataSource
=== RUN   TestAccAppgateTrustedCertificateDataSource
=== PAUSE TestAccAppgateTrustedCertificateDataSource
=== RUN   TestResourceExampleInstanceStateUpgradeV0
=== RUN   TestResourceExampleInstanceStateUpgradeV0/missing_state
=== RUN   TestResourceExampleInstanceStateUpgradeV0/move_device_limit_per_user_to_root_level
=== RUN   TestResourceExampleInstanceStateUpgradeV0/5.4_do_nothing
--- PASS: TestResourceExampleInstanceStateUpgradeV0 (0.00s)
    --- PASS: TestResourceExampleInstanceStateUpgradeV0/missing_state (0.00s)
    --- PASS: TestResourceExampleInstanceStateUpgradeV0/move_device_limit_per_user_to_root_level (0.00s)
    --- PASS: TestResourceExampleInstanceStateUpgradeV0/5.4_do_nothing (0.00s)
=== RUN   TestProvider
--- PASS: TestProvider (0.01s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccadministrativeRoleBasic
=== PAUSE TestAccadministrativeRoleBasic
=== RUN   TestAccadministrativeRoleWithScope
=== PAUSE TestAccadministrativeRoleWithScope
=== RUN   TestAccadministrativeMultiplePrivilegesValidation
=== PAUSE TestAccadministrativeMultiplePrivilegesValidation
=== RUN   TestAccadministrativeMultiplePrivilegesValidation52
=== PAUSE TestAccadministrativeMultiplePrivilegesValidation52
=== RUN   TestAccApplianceCustomizationBasic
=== PAUSE TestAccApplianceCustomizationBasic
=== RUN   TestAccApplianceBasicController
--- PASS: TestAccApplianceBasicController (12.01s)
=== RUN   TestAccApplianceConnector
=== PAUSE TestAccApplianceConnector
=== RUN   TestAccApplianceBasicGateway
=== PAUSE TestAccApplianceBasicGateway
=== RUN   TestAccApplianceBasicControllerWithoutOverrideSPA
=== PAUSE TestAccApplianceBasicControllerWithoutOverrideSPA
=== RUN   TestAccApplianceBasicControllerOverriderSPADisabled
=== PAUSE TestAccApplianceBasicControllerOverriderSPADisabled
=== RUN   TestAccAppliancePortalSetup
=== PAUSE TestAccAppliancePortalSetup
=== RUN   TestAccApplianceAdminInterfaceAddRemove
=== PAUSE TestAccApplianceAdminInterfaceAddRemove
=== RUN   TestAccApplianceLogServerFunction
--- PASS: TestAccApplianceLogServerFunction (8.27s)
=== RUN   TestAccApplianceLogForwarderElastic55
=== PAUSE TestAccApplianceLogForwarderElastic55
=== RUN   TestAccBlacklistUserBasic
=== PAUSE TestAccBlacklistUserBasic
=== RUN   TestAccClientConnectionsBasic
--- PASS: TestAccClientConnectionsBasic (2.58s)
=== RUN   TestAccClientProfileBasic
--- PASS: TestAccClientProfileBasic (13.18s)
=== RUN   TestAccConditionBasic
=== PAUSE TestAccConditionBasic
=== RUN   TestAccCriteriaScriptBasic
=== PAUSE TestAccCriteriaScriptBasic
=== RUN   TestAccDeviceScriptBasic
=== PAUSE TestAccDeviceScriptBasic
=== RUN   TestAccEntitlementScriptBasic
=== PAUSE TestAccEntitlementScriptBasic
=== RUN   TestAccEntitlementBasicPing
=== PAUSE TestAccEntitlementBasicPing
=== RUN   TestAccEntitlementBasicWithMonitor
=== PAUSE TestAccEntitlementBasicWithMonitor
=== RUN   TestAccEntitlementUpdateActionOrder
=== PAUSE TestAccEntitlementUpdateActionOrder
=== RUN   TestAccEntitlementUpdateActionHostOrder
=== PAUSE TestAccEntitlementUpdateActionHostOrder
=== RUN   TestAccEntitlementUpdateActionPortsSetOrder
=== PAUSE TestAccEntitlementUpdateActionPortsSetOrder
=== RUN   TestAccGlobalSettingsBasic
--- PASS: TestAccGlobalSettingsBasic (2.72s)
=== RUN   TestAccGlobalSettings54ProfileHostname
=== PAUSE TestAccGlobalSettings54ProfileHostname
=== RUN   TestAccConnectorIdentityProviderBasic
--- PASS: TestAccConnectorIdentityProviderBasic (3.91s)
=== RUN   TestAccLdapCertificateIdentityProvidervBasic
=== PAUSE TestAccLdapCertificateIdentityProvidervBasic
=== RUN   TestAccLdapCertificateIdentityProvidervBasic55OrGreater
=== PAUSE TestAccLdapCertificateIdentityProvidervBasic55OrGreater
=== RUN   TestAccLdapIdentityProviderBasic
=== PAUSE TestAccLdapIdentityProviderBasic
=== RUN   TestAccLdapIdentityProviderBasic55OrGreater
=== PAUSE TestAccLdapIdentityProviderBasic55OrGreater
=== RUN   TestAccLocalDatabaseIdentityProviderBasic
--- PASS: TestAccLocalDatabaseIdentityProviderBasic (3.95s)
=== RUN   TestAccRadiusIdentityProviderBasic
=== PAUSE TestAccRadiusIdentityProviderBasic
=== RUN   TestAccRadiusIdentityProviderBasic55OrGreater
=== PAUSE TestAccRadiusIdentityProviderBasic55OrGreater
=== RUN   TestAccSamlIdentityProviderBasic
=== PAUSE TestAccSamlIdentityProviderBasic
=== RUN   TestAccSamlIdentityProviderBasic55OrGreater
=== PAUSE TestAccSamlIdentityProviderBasic55OrGreater
=== RUN   TestAccIPPoolBasic
=== PAUSE TestAccIPPoolBasic
=== RUN   TestAccIPPoolV6
=== PAUSE TestAccIPPoolV6
=== RUN   TestAccLocalUserBasic
=== PAUSE TestAccLocalUserBasic
=== RUN   TestAccMfaProviderBasic
=== PAUSE TestAccMfaProviderBasic
=== RUN   TestAccAdminMfaSettingsBasic
=== PAUSE TestAccAdminMfaSettingsBasic
=== RUN   TestAccPolicyBasic
=== PAUSE TestAccPolicyBasic
=== RUN   TestAccPolicyClientSettings55
=== PAUSE TestAccPolicyClientSettings55
=== RUN   TestAccPolicyDnsSettings55
=== PAUSE TestAccPolicyDnsSettings55
=== RUN   TestAccRingfenceRuleBasicICMP
=== PAUSE TestAccRingfenceRuleBasicICMP
=== RUN   TestAccRingfenceRuleBasicTCP
=== PAUSE TestAccRingfenceRuleBasicTCP
=== RUN   TestAccSiteBasic
=== PAUSE TestAccSiteBasic
=== RUN   TestAccSiteBasicAwsResolverWithoutSecret
=== PAUSE TestAccSiteBasicAwsResolverWithoutSecret
=== RUN   TestAccSite55Attributes
=== PAUSE TestAccSite55Attributes
=== RUN   TestAccTrustedCertificateBasic
=== PAUSE TestAccTrustedCertificateBasic
=== CONT  TestAccAppgateAdministrativeRoleDataSource
=== CONT  TestAccLdapIdentityProviderBasic
=== CONT  TestAccApplianceLogForwarderElastic55
=== CONT  TestAccTrustedCertificateBasic
=== CONT  TestAccLdapCertificateIdentityProvidervBasic55OrGreater
=== CONT  TestAccEntitlementBasicWithMonitor
=== CONT  TestAccEntitlementBasicPing
=== CONT  TestAccApplianceBasicControllerOverriderSPADisabled
=== CONT  TestAccLdapCertificateIdentityProvidervBasic
=== CONT  TestAccApplianceAdminInterfaceAddRemove
=== CONT  TestAccGlobalSettings54ProfileHostname
=== CONT  TestAccApplianceConnector
=== CONT  TestAccAppliancePortalSetup
=== CONT  TestAccApplianceCustomizationBasic
=== CONT  TestAccApplianceBasicControllerWithoutOverrideSPA
=== CONT  TestAccApplianceBasicGateway
=== CONT  TestAccadministrativeMultiplePrivilegesValidation52
=== CONT  TestAccEntitlementScriptBasic
=== CONT  TestAccEntitlementUpdateActionPortsSetOrder
=== CONT  TestAccadministrativeMultiplePrivilegesValidation
=== CONT  TestAccApplianceLogForwarderElastic55
    resource_appgate_appliance_test.go:3010: Test only for 5.5 and above, appliance.log_forwarder_elasticseach specific testcase
=== CONT  TestAccAppliancePortalSetup
    resource_appgate_appliance_test.go:2170: Test only for 5.5 and above, appliance.portal is only supported in 5.4 and above.
=== CONT  TestAccLdapCertificateIdentityProvidervBasic55OrGreater
    resource_appgate_identity_provider_ldap_certificate_test.go:293: Test only for 5.5 and above, on_boarding_two_factor.0.device_limit_per_user updated behaviour in > 5.5
=== CONT  TestAccadministrativeMultiplePrivilegesValidation52
    resource_appgate_administrative_role_test.go:272: Test is only for 5.2, privileges.target OnBoardedDevice
--- SKIP: TestAccApplianceLogForwarderElastic55 (1.15s)
=== CONT  TestAccDeviceScriptBasic
--- SKIP: TestAccLdapCertificateIdentityProvidervBasic55OrGreater (1.29s)
=== CONT  TestAccCriteriaScriptBasic
--- SKIP: TestAccAppliancePortalSetup (1.50s)
=== CONT  TestAccConditionBasic
--- SKIP: TestAccadministrativeMultiplePrivilegesValidation52 (1.62s)
=== CONT  TestAccBlacklistUserBasic
--- PASS: TestAccAppgateAdministrativeRoleDataSource (7.58s)
=== CONT  TestAccEntitlementUpdateActionHostOrder
--- PASS: TestAccGlobalSettings54ProfileHostname (8.29s)
=== CONT  TestAccAdminMfaSettingsBasic
--- PASS: TestAccBlacklistUserBasic (6.67s)
=== CONT  TestAccSite55Attributes
--- PASS: TestAccadministrativeMultiplePrivilegesValidation (8.35s)
=== CONT  TestAccSiteBasicAwsResolverWithoutSecret
--- PASS: TestAccTrustedCertificateBasic (8.58s)
=== CONT  TestAccSiteBasic
--- PASS: TestAccDeviceScriptBasic (7.51s)
=== CONT  TestAccRingfenceRuleBasicTCP
--- PASS: TestAccEntitlementScriptBasic (8.81s)
=== CONT  TestAccRingfenceRuleBasicICMP
=== CONT  TestAccSite55Attributes
    resource_appgate_site_test.go:717: Test only for 5.5 and above, dns_forwarding only supported in > 5.5
--- PASS: TestAccCriteriaScriptBasic (7.94s)
=== CONT  TestAccPolicyDnsSettings55
--- SKIP: TestAccSite55Attributes (1.09s)
=== CONT  TestAccPolicyClientSettings55
--- PASS: TestAccConditionBasic (8.05s)
=== CONT  TestAccPolicyBasic
=== CONT  TestAccPolicyDnsSettings55
    resource_appgate_policy_test.go:320: Test only for 5.5 and above, appliance.portal is only supported in 5.4 and above.
--- PASS: TestAccEntitlementBasicPing (10.19s)
=== CONT  TestAccAppgateMfaProviderDataSource
--- PASS: TestAccApplianceConnector (10.28s)
=== CONT  TestAccadministrativeRoleWithScope
--- SKIP: TestAccPolicyDnsSettings55 (1.21s)
=== CONT  TestAccadministrativeRoleBasic
=== CONT  TestAccPolicyClientSettings55
    resource_appgate_policy_test.go:146: Test only for 5.5 and above, appliance.portal is only supported in 5.4 and above.
--- SKIP: TestAccPolicyClientSettings55 (1.37s)
=== CONT  TestAccAppgateTrustedCertificateDataSource
--- PASS: TestAccLdapCertificateIdentityProvidervBasic (11.23s)
=== CONT  TestAccSamlIdentityProviderBasic55OrGreater
--- PASS: TestAccApplianceBasicControllerOverriderSPADisabled (11.25s)
=== CONT  TestAccMfaProviderBasic
--- PASS: TestAccApplianceBasicControllerWithoutOverrideSPA (11.33s)
=== CONT  TestAccLocalUserBasic
--- PASS: TestAccLdapIdentityProviderBasic (11.64s)
=== CONT  TestAccIPPoolV6
--- PASS: TestAccApplianceBasicGateway (11.65s)
=== CONT  TestAccIPPoolBasic
=== CONT  TestAccSamlIdentityProviderBasic55OrGreater
    resource_appgate_identity_provider_saml_test.go:857: Test only for 5.5 and above, on_boarding_two_factor.0.device_limit_per_user updated behaviour in > 5.5
--- SKIP: TestAccSamlIdentityProviderBasic55OrGreater (1.05s)
=== CONT  TestAccRadiusIdentityProviderBasic55OrGreater
--- PASS: TestAccApplianceCustomizationBasic (13.31s)
=== CONT  TestAccSamlIdentityProviderBasic
=== CONT  TestAccRadiusIdentityProviderBasic55OrGreater
    resource_appgate_identity_provider_radius_test.go:265: Test only for 5.5 and above, on_boarding_two_factor.0.device_limit_per_user updated behaviour in > 5.5
--- SKIP: TestAccRadiusIdentityProviderBasic55OrGreater (1.32s)
=== CONT  TestAccRadiusIdentityProviderBasic
--- PASS: TestAccAdminMfaSettingsBasic (7.66s)
=== CONT  TestAccAppgateApplianceCustomizationDataSource
--- PASS: TestAccApplianceAdminInterfaceAddRemove (16.25s)
=== CONT  TestAccLdapIdentityProviderBasic55OrGreater
--- PASS: TestAccRingfenceRuleBasicICMP (7.80s)
=== CONT  TestAccEntitlementUpdateActionOrder
--- PASS: TestAccAppgateMfaProviderDataSource (6.59s)
--- PASS: TestAccSiteBasicAwsResolverWithoutSecret (8.52s)
--- PASS: TestAccRingfenceRuleBasicTCP (8.30s)
=== CONT  TestAccLdapIdentityProviderBasic55OrGreater
    resource_appgate_identity_provider_ldap_test.go:257: Test only for 5.5 and above, on_boarding_two_factor.0.device_limit_per_user updated behaviour in > 5.5
--- SKIP: TestAccLdapIdentityProviderBasic55OrGreater (1.16s)
--- PASS: TestAccAppgateTrustedCertificateDataSource (6.74s)
--- PASS: TestAccPolicyBasic (8.01s)
--- PASS: TestAccadministrativeRoleBasic (8.08s)
--- PASS: TestAccLocalUserBasic (7.23s)
--- PASS: TestAccMfaProviderBasic (7.31s)
--- PASS: TestAccEntitlementBasicWithMonitor (18.58s)
--- PASS: TestAccadministrativeRoleWithScope (8.33s)
--- PASS: TestAccIPPoolV6 (6.97s)
--- PASS: TestAccIPPoolBasic (6.97s)
--- PASS: TestAccEntitlementUpdateActionPortsSetOrder (18.63s)
--- PASS: TestAccAppgateApplianceCustomizationDataSource (4.85s)
--- PASS: TestAccRadiusIdentityProviderBasic (8.00s)
--- PASS: TestAccEntitlementUpdateActionHostOrder (14.89s)
--- PASS: TestAccEntitlementUpdateActionOrder (10.36s)
--- PASS: TestAccSamlIdentityProviderBasic (18.78s)
--- PASS: TestAccSiteBasic (23.56s)
PASS
ok  	github.com/appgate/terraform-provider-appgatesdp/appgate	116.742s
```
</details>

### CODE REVIEW

File: appgate/resource_appgate_appliance_controller.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		CreateContext: resourceAppgateApplianceControllerActivationCreate,
+		ReadContext:   resourceAppgateApplianceControllerActivationRead,
+		UpdateContext: resourceAppgateApplianceControllerActivationUpdate,
+		DeleteContext: resourceAppgateApplianceControllerActivationDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(15 * time.Minute),
+			Update: schema.DefaultTimeout(15 * time.Minute),
+			Delete: schema.DefaultTimeout(15 * time.Minute),
+		},
```
Comment (thomascellerier) [inline appgate/resource_appgate_appliance_controller.go:29]: Assuming these timeouts mean the new controller is fully joined (the database was fully synced) and operational. Enabling the controller role can take longer than this depending on the database size and latency. E.g. if you have controllers on the opposite side of the world with a fairly big database. For most cases it is more than enough though.

These are just defaults right?

File: appgate/resource_appgate_appliance_controller.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		appliance.SetAdminInterface(ainterface)
+	}
+
+	retryErr := resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		_, _, err := api.AppliancesIdPut(ctx, id).Appliance(appliance).Authorization(token).Execute()
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		// wait a few seconds and make sure the controller react to the changes.
+		time.Sleep(time.Duration(10) * time.Second)
```
Comment (thomascellerier) [inline appgate/resource_appgate_appliance_controller.go:104]: Is a sleep really required here? Can't `waitForController` be able to handle it regardless?

File: website/docs/guides/ha_controllers.markdown [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+sidebar_current: "docs-appgatesdp-guide-high_availability_controllers"
+description: |-
+  Provisioning high availability controllers
+---
+
+## Provisioning high availability controllers
+
+
+In appliance >= 5.4.0 we cant create more then 1 controller at the time,
+it means that we must first create an appliance, seed it, then activate the controller function.
```
Comment (thomascellerier) [inline website/docs/guides/ha_controllers.markdown:13]: Just to clear things up, you never could do that as joining/leaving the cluster can only be done 1 node at a time.
It was allowed by the API but would result in errors.

**Review by thomascellerier — APPROVED** [review]
Looks good overall, just some small comments.
Wondering if we can remove the use of `time.Sleep` as well, or at least move it to the `waitForControllers` function.

File: appgate/resource_appgate_appliance_controller.go (hunk 1 above)
Comment (danijeel) [reply appgate/resource_appgate_appliance_controller.go:29]: yes, this is default values can take be overwritten in the resource.
https://www.terraform.io/docs/language/resources/syntax.html#operation-timeouts
for example
```hcl
resource "appgatesdp_appliance_controller_activation" "activate_second_controller" {
  timeouts {
    create = "60m"
    delete = "2h"
  }
  appliance_id = "69723c1d-0b5f-44f8-9f29-3eed57618440"
  controller {
    enabled = true
  }
  admin_interface {
    hostname   = "envy-10-97-170-3.devops"
    https_port = 8443
    https_ciphers = [
      "ECDHE-RSA-AES256-GCM-SHA384",
      "ECDHE-RSA-AES128-GCM-SHA256"
    ]
  }
}

```

**Pull request merged**